src/repository.rs

```diff
@@ -19,53 +23,142 @@ pub use self::http::{HttpRepository, HttpRepositoryBuilder};
+    ///
+    /// [extension]: crate::interchange::DataInterchange::extension
+    fn store_metadata<'a, R, D>(
         &'a self,
         meta_path: &'a MetadataPath,
         version: &'a MetadataVersion,
-        max_length: Option<usize>,
-        hash_data: Option<(&'static HashAlgorithm, HashValue)>,
-    ) -> BoxFuture<'a, Result<SignedMetadata<D, M>>>
+        metadata: R,
+    ) -> BoxFuture<'a, Result<()>>
     where
-        M: Metadata + 'static;
+        R: AsyncRead + Send + Unpin + 'a,
+        D: DataInterchange + Sync;
 
-    /// Store the given target.
+    /// Store the provided `target` in a location identified by `target_path`, overwriting any
+    /// existing target at that location.
     fn store_target<'a, R>(
         &'a self,
-        read: R,
+        target: R,
         target_path: &'a TargetPath,
     ) -> BoxFuture<'a, Result<()>>
     where
         R: AsyncRead + Send + Unpin + 'a;
+}
+
+impl<T> RepositoryProvider for &T
+where
+    T: RepositoryProvider,
+{
+    fn fetch_metadata<'a, D>(
+        &'a self,
+        meta_path: &'a MetadataPath,
+        version: &'a MetadataVersion,
+        max_length: Option<usize>,
+        hash_data: Option<(&'static HashAlgorithm, HashValue)>,
+    ) -> BoxFuture<'a, Result<Box<dyn AsyncRead + Send + Unpin>>>
+    where
+        D: DataInterchange + Sync,
+    {
+        (**self).fetch_metadata::<D>(meta_path, version, max_length, hash_data)
+    }
 
-    /// Fetch the given target.
     fn fetch_target<'a>(
         &'a self,
         target_path: &'a TargetPath,
         target_description: &'a TargetDescription,
-    ) -> BoxFuture<'a, Result<Box<dyn AsyncRead + Send + Unpin>>>;
+    ) -> BoxFuture<'a, Result<Box<dyn AsyncRead + Send + Unpin>>> {
+        (**self).fetch_target(target_path, target_description)
+    }
+}
+
+impl<T> RepositoryStorage for &T
+where
+    T: RepositoryStorage,
+{
+    fn store_metadata<'a, R, D>(
+        &'a self,
+        meta_path: &'a MetadataPath,
+        version: &'a MetadataVersion,
+        metadata: R,
+    ) -> BoxFuture<'a, Result<()>>
+    where
+        R: AsyncRead + Send + Unpin + 'a,
+        D: DataInterchange + Sync,
+    {
+        (**self).store_metadata::<_, D>(meta_path, version, metadata)
+    }
 
-    /// Perform a sanity check that `M`, `Role`, and `MetadataPath` all desrcribe the same entity.
+    fn store_target<'a, R>(
+        &'a self,
+        target: R,
+        target_path: &'a TargetPath,
+    ) -> BoxFuture<'a, Result<()>>
+    where
+        R: AsyncRead + Send + Unpin + 'a,
+    {
+        (**self).store_target(target, target_path)
+    }
+}
+
+/// A wrapper around an implementation of [`RepositoryProvider`] and/or [`RepositoryStorage`] tied
+/// to a specific [`DataInterchange`](crate::interchange::DataInterchange) that will enforce
+/// provided length limits and hash checks.
+#[derive(Debug, Clone)]
+pub struct Repository<R, D> {
```

Comment:
I think we could make this `pub(crate)`. It's mainly about making sure the client doesn't accidentally access the repositories directly. I'm not sure if external users would need this.

Reply:
Fine with me.  It was a little more convenient to use in the test binaries in this crate, but it isn't too terrible to use the repository implementations directly, and I suppose I could add some similar methods to EphemeralRepository to simplify testing if we find ourselves doing a lot of repository manipulation from external tests.